jormungandr/src/blockchain/storage.rs

```diff
@@ -37,205 +113,306 @@ struct BlockIterState {
     pending_infos: Vec<BlockInfo<HeaderHash>>,
 }
 
-impl Storage {
+impl Storage03 {
     pub fn new(storage: NodeStorage) -> Self {
-        Storage {
-            read_connection: storage.clone(),
-            write_connection_lock: Lock::new(storage),
-        }
+        let mut rt = runtime::Builder::new()
+            .name_prefix("new-storage-worker-")
+            .core_threads(1)
+            .build()
+            .unwrap();
+
+        rt.block_on_std(async move {
+            let manager = ConnectionManager::new(storage);
+            let pool = Pool::builder().build(manager).await.unwrap();
+            let write_lock = Arc::new(Semaphore::new(1));
```

Comment:
What's the benefit of semaphore over `Mutex` here?

Reply:
It's about lifetimes. Mutex guards depend on the lifetime of a mutex, which needs to be `'static` in this case (because of the requirements of `spawn_blocking`).

Reply:
But the result of `Semaphore::acquire` is lifetime-bound too?

Reply:
It is, but is not used by any of closures running in `spawn_blocking`.

Reply:
I still don't get how is this different from replacing all occurrences of `Semaphore` with the `Mutex` equivalent, except semaphores being slower.

Reply:
Do you mean `Mutex<()>`?

Reply:
Yes, as you obviously just need a critical section. Semaphores add an overhead of keeping a counter, which we don't need.